src/canopy/llm/openai.py

```diff
@@ -206,3 +207,135 @@ async def agenerate_queries(self,
+
+        To read more about this feature, see: https://platform.openai.com/docs/guides/gpt/function-calling
+
+        Note: this function is wrapped in a retry decorator to handle transient errors.
+
+        Args:
+            messages: Messages (chat history) to send to the model.
+            function: Function to call. See canopy.llm.models.Function for more details.
+            max_tokens: Maximum number of tokens to generate. Defaults to None (generates until stop sequence or until hitting max context size).
+            model_params: Model parameters to use for this request. Defaults to None (uses the default model parameters).
+                          Overrides the default model parameters if set on initialization.
+                          For example, you can pass: {"temperature": 0.9, "top_p": 1.0} to override the default temperature and top_p.
+                          see: https://platform.openai.com/docs/api-reference/chat/create
+
+        Returns:
+            dict: Function call arguments as a dictionary.
+
+        Usage:
+            >>> from canopy.llm import AzureOpenAILLM
+            >>> from canopy.llm.models import Function, FunctionParameters, FunctionArrayProperty
+            >>> from canopy.models.data_models import UserMessage
+            >>> llm = AzureOpenAILLM()
+            >>> messages = [UserMessage(content="I was wondering what is the capital of France?")]
+            >>> function = Function(
+            ...     name="query_knowledgebase",
+            ...     description="Query search engine for relevant information",
+            ...     parameters=FunctionParameters(
+            ...         required_properties=[
+            ...             FunctionArrayProperty(
+            ...                 name="queries",
+            ...                 items_type="string",
+            ...                 description='List of queries to send to the search engine.',
+            ...             ),
+            ...         ]
+            ...     )
+            ... )
+            >>> llm.enforced_function_call(messages, function)
+            {'queries': ['capital of France']}
+        """  # noqa: E501
+
+        model_params_dict: Dict[str, Any] = deepcopy(self.default_model_params)
+        model_params_dict.update(model_params or {})
+
+        function_dict = cast(ChatCompletionToolParam, function.dict())
+
+        chat_completion = self._client.chat.completions.create(
+            model=self.model_name,
+            messages=[m.dict() for m in messages],
+            functions=[function_dict],
+            function_call={"name": function.name},
+            max_tokens=max_tokens,
+            **model_params_dict,
+        )
+
+        result = chat_completion.choices[0].message.function_call.arguments
+        arguments = json.loads(result)
+
+        jsonschema.validate(instance=arguments, schema=function.parameters.dict())
+        return arguments
+
```

Comment:
Note that the `.available_models` property is broken, as it seems that Azure OpenAI API doesn't allow listing all of your deployments (==models).  
We should wrap that with a clear error message, and point to where the list of available deployments can be found in the Azure console